File: tools/resize_icon.py
"""
Resize app icon to 512x512, content inside 312px diameter circle, no alpha.
"""
from PIL import Image
import math

OUTPUT_SIZE = 512
CIRCLE_DIAMETER = 312
GREEN = (0x2E, 0x7D, 0x32)  # #2E7D32

def main():
    src = Image.open("assets/icons/app_icon_1024.png").convert("RGBA")
    w, h = src.size

    # Scale source so that it fits in CIRCLE_DIAMETER (content circle)
    scaled = src.resize((CIRCLE_DIAMETER, CIRCLE_DIAMETER), Image.Resampling.LANCZOS)

    # Create 512x512 canvas with solid green (no alpha)
    out = Image.new("RGB", (OUTPUT_SIZE, OUTPUT_SIZE), GREEN)

    # Circular mask: 1 inside circle, 0 outside
    mask = Image.new("L", (CIRCLE_DIAMETER, CIRCLE_DIAMETER), 0)
    center = CIRCLE_DIAMETER / 2.0
    radius = CIRCLE_DIAMETER / 2.0
    for y in range(CIRCLE_DIAMETER):
        for x in range(CIRCLE_DIAMETER):
            if (x - center) ** 2 + (y - center) ** 2 <= radius ** 2:
                mask.putpixel((x, y), 255)

    # Composite scaled image onto green background using circle mask, then paste onto out
    # First: composite scaled (RGBA) onto green within the circle
    box_bg = Image.new("RGB", (CIRCLE_DIAMETER, CIRCLE_DIAMETER), GREEN)
    box_bg.paste(scaled, (0, 0), scaled)  # paste with alpha from scaled
    # Now box_bg is RGB (no alpha) for the circle content. Apply mask: where mask=0 use green.
    paste_x = (OUTPUT_SIZE - CIRCLE_DIAMETER) // 2  # 100
    paste_y = paste_x
    for y in range(CIRCLE_DIAMETER):
        for x in range(CIRCLE_DIAMETER):
            if mask.getpixel((x, y)) == 255:
                out.putpixel((paste_x + x, paste_y + y), box_bg.getpixel((x, y)))
            # else keep out's pixel (already green)

    out.save("assets/icons/app_icon_512.png", "PNG")
    print("Saved assets/icons/app_icon_512.png (512x512, content in 312px circle, no alpha)")

if __name__ == "__main__":
    main()

File: bin/resize_icon.dart
// Resize app icon: 512x512, content in 312px diameter circle, no alpha.
// Run from project root: dart run bin/resize_icon.dart

import 'dart:io';

import 'package:image/image.dart' as img;

void main() {
  const outputSize = 512;
  const circleDiameter = 312;
  const greenR = 0x2E, greenG = 0x7D, greenB = 0x32;

  final bytes = File('assets/icons/app_icon_1024.png').readAsBytesSync();
  final src = img.decodeImage(bytes);
  if (src == null) {
    print('Failed to decode assets/icons/app_icon_1024.png');
    exit(1);
  }

  // Resize source to fit inside circle (312x312)
  final scaled = img.copyResize(
    src,
    width: circleDiameter,
    height: circleDiameter,
    interpolation: img.Interpolation.linear,
  );

  // Use scaled as-is; getPixel gives r,g,b,a (a=255 if 3-channel)
  final scaledRgba = scaled;

  // Output: 512x512, no transparency (all opaque)
  final out = img.Image(width: outputSize, height: outputSize, numChannels: 4);
  final center = circleDiameter / 2.0;
  final radius = circleDiameter / 2.0;
  const pasteOffset = (outputSize - circleDiameter) ~/ 2; // 100

  for (var oy = 0; oy < outputSize; oy++) {
    for (var ox = 0; ox < outputSize; ox++) {
      final sx = ox - pasteOffset;
      final sy = oy - pasteOffset;
      final inCircle = sx >= 0 &&
          sx < circleDiameter &&
          sy >= 0 &&
          sy < circleDiameter &&
          (sx - center) * (sx - center) + (sy - center) * (sy - center) <=
              radius * radius;

      if (inCircle) {
        final p = scaledRgba.getPixel(sx, sy);
        final a = p.a.toInt() / 255.0;
        final r = (p.r.toInt() * a + greenR * (1 - a)).round().clamp(0, 255);
        final g = (p.g.toInt() * a + greenG * (1 - a)).round().clamp(0, 255);
        final b = (p.b.toInt() * a + greenB * (1 - a)).round().clamp(0, 255);
        out.setPixelRgba(ox, oy, r, g, b, 255);
      } else {
        out.setPixelRgba(ox, oy, greenR, greenG, greenB, 255);
      }
    }
  }

  // Copy to 3-channel image so PNG has no alpha channel (24-bit)
  final outRgb = img.Image(width: outputSize, height: outputSize, numChannels: 3);
  for (var y = 0; y < outputSize; y++) {
    for (var x = 0; x < outputSize; x++) {
      final p = out.getPixel(x, y);
      outRgb.setPixelRgba(x, y, p.r.toInt(), p.g.toInt(), p.b.toInt(), 255);
    }
  }
  final outFile = File('assets/icons/app_icon_512.png');
  outFile.writeAsBytesSync(img.encodePng(outRgb));
  print('Saved ${outFile.path} (512x512, content in ${circleDiameter}px circle, no alpha)');
}
